src/main/telemetry/msp_shared.c

```diff
@@ -0,0 +1,217 @@
+
+#include "msp/msp.h"
+
+#include "rx/crsf.h"
+#include "rx/msp.h"
+
+#include "telemetry/msp_shared.h"
+#include "telemetry/smartport.h"
+
+#define TELEMETRY_MSP_VERSION    1
+#define TELEMETRY_MSP_VER_SHIFT  5
+#define TELEMETRY_MSP_VER_MASK   (0x7 << TELEMETRY_MSP_VER_SHIFT)
+#define TELEMETRY_MSP_ERROR_FLAG (1 << 5)
+#define TELEMETRY_MSP_START_FLAG (1 << 4)
+#define TELEMETRY_MSP_SEQ_MASK   0x0F
+#define TELEMETRY_MSP_RES_ERROR (-10)
+
+enum {
+    TELEMETRY_MSP_VER_MISMATCH=0,
+    TELEMETRY_MSP_CRC_ERROR=1,
+    TELEMETRY_MSP_ERROR=2
+};
+
+#define REQUEST_BUFFER_SIZE 64
+#define RESPONSE_BUFFER_SIZE 64
+    
+STATIC_UNIT_TESTED uint8_t checksum = 0;
+STATIC_UNIT_TESTED mspPackage_t mspPackage;
+static mspRxBuffer_t mspRxBuffer;
+static mspTxBuffer_t mspTxBuffer;
+static mspPacket_t mspRxPacket;
+static mspPacket_t mspTxPacket;
+
+void initSharedMsp() {
+    mspPackage.requestBuffer = (uint8_t *)&mspRxBuffer;
+    mspPackage.requestPacket = &mspRxPacket;
+    mspPackage.requestPacket->buf.ptr = mspPackage.requestBuffer;
+    mspPackage.requestPacket->buf.end = mspPackage.requestBuffer;
+
+    mspPackage.responseBuffer = (uint8_t *)&mspTxBuffer;
+    mspPackage.responsePacket = &mspTxPacket;
+    mspPackage.responsePacket->buf.ptr = mspPackage.responseBuffer;
+    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+}
+
+static void processMspPacket()
+{
+    mspPackage.responsePacket->cmd = 0;
+    mspPackage.responsePacket->result = 0;
+    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+
+    mspPostProcessFnPtr mspPostProcessFn = NULL;
+    if (mspFcProcessCommand(mspPackage.requestPacket, mspPackage.responsePacket, &mspPostProcessFn) == MSP_RESULT_ERROR) {
+        sbufWriteU8(&mspPackage.responsePacket->buf, TELEMETRY_MSP_ERROR);
+    }
+    if (mspPostProcessFn) {
+        mspPostProcessFn(NULL);
+    }
+
+    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+}
+
+void sendMspErrorResponse(uint8_t error, int16_t cmd)
+{
+    mspPackage.responsePacket->cmd = cmd;
+    mspPackage.responsePacket->result = 0;    
+    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+
+    sbufWriteU8(&mspPackage.responsePacket->buf, error);
+    mspPackage.responsePacket->result = TELEMETRY_MSP_RES_ERROR;
+    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+}
+
+bool handleMspFrame(uint8_t *frameStart, uint8_t *frameEnd)
+{
+    static uint8_t mspStarted = 0;
+    static uint8_t lastSeq = 0;
+
+    mspPacket_t *packet = mspPackage.requestPacket;
+    sbuf_t *frameBuf = &mspPackage.requestFrame;
+
+    if (sbufBytesRemaining(&mspPackage.responsePacket->buf) > 0) {
+        return false;
+    } 
+      
+    if (mspStarted == 0) {
+        initSharedMsp();
+    }
+    
+    uint8_t frameLength = frameEnd - frameStart;
+    uint8_t frame[frameLength];
+
+    memcpy(&frame, frameStart, frameLength);
+
+    frameBuf->ptr = (uint8_t*)&frame;
```

Comment:
Use `sbuf_t *frameBuf = (&mspPackage.requestFrame, frameStart, frameEnd);`

Reply:
I'm assuming you mean `sbufInit(&mspPackage...`

Coming soon :)

Reply:
only problem with using frameStart, frameEnd is that incoming, unwanted request data will pollute the data stored at these pointers.  This was the inspiration for the memcpy.

Reply:
Seems to me that `sbuf_t *frameBuf = sbufInit(&mspPackage.requestFrame, &frame, &frame + frameLength);` would be more memory safe.  Open to suggestions.